src/clims/services/container.py

```diff
@@ -1,13 +1,99 @@
 from __future__ import absolute_import
 
+import re
+import six
 from clims.services.extensible import ExtensibleBase
 from clims.models import Container, ContainerVersion
 
 
+class IndexOutOfBounds(Exception):
+    pass
+
+
+class ContainerIndex(object):
+    """
+    Represent an index into a regular Container.
+    """
+
+    def __init__(self, x, y, z):
+        self.x = x
+        self.y = y
+        self.z = z
+
+    @property
+    def raw(self):
+        # Returns the tuple that's used under the hood when saving the item.
+        return self.x, self.y, self.z
+
+    @classmethod
+    def from_string(s):
+        raise NotImplementedError("It's not possible to use this index type with a string")
+
+    @classmethod
+    def from_any_type(cls, key):
+        # Creates an index for the input key if possible.
+        if isinstance(key, six.binary_type):
+            return cls.from_string(key)
+        if isinstance(key, tuple):
+            return cls(*key)
+        else:
+            raise NotImplementedError("Can't use {} as an index".format(type(key)))
+
+
+class InvalidContainerIndex(Exception):
+    pass
+
+
+class PlateIndex(ContainerIndex):
+    """
+    Wraps an index to a plate from the generic zero based (x, y, z) coordinates used in the DB.
+
+    Indexes into a plate are 0-based (row, column) tuples.
+
+    They can also use the alternative <row-letter>[:]<column-number>, where row-letter is e.g. A for
+    row 0 and column-number is a one-based column number. This allows users to address the top-left
+    most location of the plate with 'A:1'
+    """
+
+    STRING_PATTERN = re.compile(r'(?P<row>\w):?(?P<col>\d+)')
+
+    def __init__(self, row, column):
+        super(PlateIndex, self).__init__(column, row, 0)
+
+    @property
+    def row(self):
+        return self.y
+
+    @property
+    def column(self):
+        return self.x
+
+    @classmethod
+    def from_string(cls, key):
+        """
+        Given a human-readable index into the container, e.g. "A:1", creates a new index.
+        """
+        m = cls.STRING_PATTERN.match(key)
+        if not m:
+            raise InvalidContainerIndex(
+                "Can't interpret '{}' as a container location".format(key))
+        row = m.group('row').upper()
```

Comment:
Minor thing, but I don't know if we want to validate that the letter is in the A-Z range? Or is ":hear_no_evil: :1" a valid index?